pkg/spanconfig/spanconfigsqlwatcher/buffer.go

```diff
@@ -119,27 +185,94 @@ func (b *buffer) flushEvents(
 // combined frontier timestamp is also returned.
 func (b *buffer) flush(
 	ctx context.Context,
-) (updates []spanconfig.DescriptorUpdate, _ hlc.Timestamp, _ error) {
+) (sqlUpdates []spanconfig.SQLUpdate, _ hlc.Timestamp, _ error) {
 	events, combinedFrontierTS := b.flushEvents(ctx)
-	sort.Slice(events, func(i, j int) bool {
-		ei, ej := events[i].(event), events[j].(event)
-		if ei.update.ID == ej.update.ID {
-			return ei.timestamp.Less(ej.timestamp)
-		}
-		return ei.update.ID < ej.update.ID
+	bufferedEvents := make(bufferedEvents, 0, len(events))
+	for _, ev := range events {
+		bufferedEvents = append(bufferedEvents, ev.(event))
+	}
+	// Nil out the underlying slice since we have copied over the events.
+	events = nil
+
+	// The events slice can contain a SQLUpdate that applies to a descriptor, or a
+	// SQLUpdate that applies to a protected timestamp record. We sort the
+	// protected timestamp SQLUpdates to the end of the slice, so that we can
+	// perform the subsequent deduplication of SQLUpdates. We do this instead of
+	// pre-processing events to separate the two kinds of updates, to save on the
+	// allocation of an additional `events` slices.
+	sort.Sort(bufferedEvents)
+
+	// Find the index before which all events are SQLUpdates on descriptors.
+	descriptorUpdatesIdx := sort.Search(len(bufferedEvents), func(i int) bool {
+		update := bufferedEvents[i].update
+		return update.IsProtectedTimestampUpdate()
 	})
-	for i, ev := range events {
-		if i == 0 || events[i-1].(event).update.ID != ev.(event).update.ID {
-			updates = append(updates, ev.(event).update)
+
+	// Deduplicate the SQLUpdates emitted by the buffer that apply to descriptors.
+	for i, ev := range bufferedEvents[:descriptorUpdatesIdx] {
+		update := ev.update
+		descriptorUpdate := update.GetDescriptorUpdate()
+		if i == 0 {
+			sqlUpdates = append(sqlUpdates, update)
 			continue
 		}
-		descType, err := combine(updates[len(updates)-1].DescriptorType, ev.(event).update.DescriptorType)
+
+		prevUpdate := bufferedEvents[i-1].update
+		if prevUpdate.GetDescriptorUpdate().ID != descriptorUpdate.ID {
+			sqlUpdates = append(sqlUpdates, update)
+			continue
+		}
+
+		prevDescriptorSQLUpdate := sqlUpdates[len(sqlUpdates)-1].GetDescriptorUpdate()
+		descType, err := combine(prevDescriptorSQLUpdate.Type,
+			descriptorUpdate.Type)
 		if err != nil {
 			return nil, hlc.Timestamp{}, err
 		}
-		updates[len(updates)-1].DescriptorType = descType
+		sqlUpdates[len(sqlUpdates)-1] = spanconfig.MakeDescriptorSQLUpdate(
+			prevDescriptorSQLUpdate.ID, descType)
 	}
-	return updates, combinedFrontierTS, nil
+
+	// Truncate the slice to only include ProtectedTimestampUpdates now that we
+	// have copied over all de-duped DescriptorUpdates.
+	bufferedEvents = bufferedEvents[descriptorUpdatesIdx:]
+
+	// Deduplicate the SQLUpdates emitted by the buffer that apply to protected
+	// timestamps.
+	for i, ev := range bufferedEvents {
+		update := ev.update
```

Comment:
Can we add a log.Fatal if prev event timestamp is > the current? Just being paranoid -- could do it when copying over elements 1-by-1.

Reply:
hmm adding it here might not be right cause we major sort on tenant ID, so you could have:

`event{ tenantTarget:1, ts:7}, event{tenantTarget:2, ts5}`

I added it when we are copying over yup.